### PR TITLE
Regression testing

### PR DESCRIPTION
**Pull request by jenny-m-wood**
Changes to pyQuARC from regression testing

### CODE REVIEW

File: pyQuARC/code/custom_validator.py [hunk 1]
```diff
@@ -179,11 +179,11 @@ def uniqueness_check(list_of_objects, key):
         seen, duplicates = set(), set()
         if isinstance(list_of_objects, list):
             for url_obj in list_of_objects:
-                if description := url_obj.get(key) in seen:
+                description = url_obj.get(key)
+                if description in seen:
```
Comment (xhagrg) [inline pyQuARC/code/custom_validator.py:183]: Why the change?

File: pyQuARC/code/gcmd_validator.py [hunk 2]
```diff
@@ -2,7 +2,7 @@
 import os
 import requests
 
-from pyQuARC.code.utils import get_headers
+from .utils import get_headers
```
Comment (xhagrg) [inline pyQuARC/code/gcmd_validator.py:5]: This might work when not using package, but does it work when using "installed" package?

File: pyQuARC/main.py [hunk 3]
```diff
@@ -6,19 +6,20 @@
 
 from tqdm import tqdm
 
-from pyQuARC.code.utils import get_headers
-
 if __name__ == '__main__':
+    from code.utils import get_headers
```
Comment (xhagrg) [inline pyQuARC/main.py:10]: why twice?

File: pyQuARC/main.py [hunk 4]
```diff
@@ -6,19 +6,20 @@
...
     from code.checker import Checker
     from code.constants import COLOR, ECHO10_C, SUPPORTED_FORMATS
     from code.downloader import Downloader
     from code.utils import get_cmr_url, is_valid_cmr_url
```
Comment (xhagrg) [inline pyQuARC/main.py:14]: add get_headers to the list of methods listed here?

File: pyQuARC/schemas/rule_mapping.json [hunk 5]
```diff
@@ -1531,20 +1533,20 @@
             "umm-c": [
                 {
                     "fields": [
-                        "Platform/Characteristics/Name",
-                        "Platform/Characteristics"
+                        "Platforms/Characteristics/Name",
+                        "Platforms/Characteristics"
```
Comment (xhagrg) [inline pyQuARC/schemas/rule_mapping.json:1537]: Was this the change throughout umm-c metadata?

File: tests/test_checker.py [hunk 6]
```diff
@@ -1,9 +1,9 @@
+import sys
+import os
```
Comment (xhagrg) [inline tests/test_checker.py:2]: import order alphabetical

File: tests/test_custom_checker.py [hunk 7]
```diff
@@ -1,10 +1,10 @@
...
+import sys
+import os
+sys.path.append(os.getcwd())
 from pyQuARC.code.custom_checker import CustomChecker
-
-from .fixtures.custom_checker import INPUT_OUTPUT
-from .common import read_test_metadata
-
+from tests.fixtures.custom_checker import INPUT_OUTPUT
+from tests.common import read_test_metadata
```
Comment (xhagrg) [inline tests/test_custom_checker.py:7]: were the tests not working without?

File: tests/test_string_validator.py [hunk 8]
```diff
@@ -1,6 +1,7 @@
+import sys
+import os
+sys.path.append(os.getcwd())
```
Comment (xhagrg) [inline tests/test_string_validator.py:3]: This doesn't seem necessary.

File: tests/test_schema_validator.py [hunk 9]
```diff
@@ -1,7 +1,7 @@
+import sys
 import os
-
+sys.path.append(os.getcwd())
```
Comment (xhagrg) [inline tests/test_schema_validator.py:3]: Do we need these? (besides `import os`)

File: tests/test_downloader.py [hunk 10]
```diff
@@ -1,6 +1,7 @@
+import sys
+import os
+sys.path.append(os.getcwd())
```
Comment (xhagrg) [inline tests/test_downloader.py:3]: required?

File: tests/test_datetime_validator.py [hunk 11]
```diff
@@ -1,8 +1,8 @@
+import sys
+import os
+sys.path.append(os.getcwd())
```
Comment (xhagrg) [inline tests/test_datetime_validator.py:3]: ?

File: tests/fixtures/checks_umm-g_master_test_file.json [hunk 12]
```diff
@@ -0,0 +1,379 @@
...
+    },
+    "validate_granule_data_format_against_collection_check": {
+        "valid": ["ICARTT", "MACPEX_TraceGas_AircraftInSitu_WB57_Data", "1", ""],
+        "invalid": ["ASCII", "MACPEX_TraceGas_AircraftInSitu_WB57_Data", "1", ""]
+    },
+    "granule_data_format_presence_check": {
+        "valid": ["Binary"],
+        "invalid": [""]
+    }
+}
```
Comment (xhagrg) [inline tests/fixtures/checks_umm-g_master_test_file.json:379]: add new line here.

File: tests/fixtures/checks_echo-g_master_test_file.json [hunk 13]
```diff
@@ -0,0 +1,360 @@
...
+    },
+    "validate_granule_data_format_against_collection_check": {
+        "valid": ["HDF5", "ATL06", "005", "ATLAS/ICESat-2 L3A Land Ice Height V005"],
+        "invalid": ["ICARTT", "ATL06", "005", "ATLAS/ICESat-2 L3A Land Ice Height V005"]
+    },
+    "granule_data_format_presence_check": {
+        "valid": ["Binary"],
+        "invalid": [""]
+    }
+}
```
Comment (xhagrg) [inline tests/fixtures/checks_echo-g_master_test_file.json:360]: same here.

File: tests/fixtures/checks_echo-g_master_test_file.json [hunk 14]
```diff
@@ -0,0 +1,360 @@
...
+        ]
+    },
+    "granule_single_date_time_check": {
+        "valid": ["2010-01-01T00:00:00Z", "2020-01-01T00:00:00Z", "neq"],
+        "invalid": ["2020-01-01T00:00:00Z", "2020-01-01T00:00:00Z", "neq"]
+    },
+    "granule_project_short_name_check": {
+        "valid": ["OLYMPEX", "gpmcplolyx", "1", "GPM Ground Validation Cloud Physics LiDAR (CPL) OLYMPEX V1"],
+        "invalid": ["GOES-R PLT", "gpmcplolyx", "1", "GPM Ground Validation Cloud Physics LiDAR (CPL) OLYMPEX V1"]
+    },
```
Comment (spa0002) [inline tests/fixtures/checks_echo-g_master_test_file.json:347]: Should be:
```

    "granule_project_short_name_check": {
        "valid": ["OLYMPEX", "GPM Ground Validation Cloud Physics LiDAR (CPL) OLYMPEX V1", "gpmcplolyx", "1"],
        "invalid": ["GOES-R PLT", "GPM Ground Validation Cloud Physics LiDAR (CPL) OLYMPEX V1", "gpmcplolyx", "1"]
    },
```
due to the order of the fields in the rule_mapping.json ->

```
    "granule_project_short_name_check": {
        "rule_name": "Granule Project Short Name Check",
        "fields_to_apply": {
            "echo-g": [
                {
                    "fields": [
                        "Granule/Campaigns/Campaign/ShortName",
                        "Granule/Collection/DataSetId",
                        "Granule/Collection/ShortName",
                        "Granule/Collection/VersionId"
                    ]
                }
            ],

```

File: pyQuARC/schemas/rule_mapping.json (hunk 5 above)
Comment (spa0002) [reply pyQuARC/schemas/rule_mapping.json:1537]: "Platforms" shows up as required in the latest UMM-C schema: https://git.earthdata.nasa.gov/projects/EMFD/repos/unified-metadata-model/browse/collection/v1.17.1/umm-c-json-schema.json#274

File: pyQuARC/code/custom_validator.py (hunk 1 above)
Comment (spa0002) [reply pyQuARC/code/custom_validator.py:183]: Using the following valid/invalid values from checks_umm-g_master_test_file.json:
```
    "url_description_uniqueness_check": {
        "valid": [
            [
                {
                    "Description": "Example description 1",
                    "Subtype": "GENERAL DOCUMENTATION",
                    "Type": "VIEW RELATED INFORMATION",
                    "URL": "https://www.nasa.gov/",
                    "URLContentType": "PublicationURL"
                },
                {
                    "Description": "Example description 2",
                    "Subtype": "PUBLICATIONS",
                    "Type": "VIEW RELATED INFORMATION",
                    "URL": "https://www.nasa.gov/",
                    "URLContentType": "PublicationURL"
                }
            ],
            "Description"],
        "invalid": [
            [
                {
                    "Description": "Example description 1",
                    "Subtype": "GENERAL DOCUMENTATION",
                    "Type": "VIEW RELATED INFORMATION",
                    "URL": "https://www.nasa.gov/",
                    "URLContentType": "PublicationURL"
                },
                {
                    "Description": "Example description 1",
                    "Subtype": "PUBLICATIONS",
                    "Type": "VIEW RELATED INFORMATION",
                    "URL": "https://www.nasa.gov/",
                    "URLContentType": "PublicationURL"
                }
            ],
            "Description"]
    },

```
Example test output for umm-g with
`                if description := url_obj.get(key) in seen:`
```
rule_name: URL Description Uniqueness Check
check_id: uniqueness_check
validator function: CustomValidator.uniqueness_check
test output:
with valid test input: {'valid': True, 'value': ''}
with invalid test input: {'valid': True, 'value': ''}
```
Example test output for umm-g with
```
                description = url_obj.get(key)
                if description in seen:

```
```
rule_name: URL Description Uniqueness Check
check_id: uniqueness_check
validator function: CustomValidator.uniqueness_check
test output:
with valid test input: {'valid': True, 'value': ''}
with invalid test input: {'valid': False, 'value': 'Example description 1'}

```

File: tests/test_custom_checker.py (hunk 7 above)
Comment (spa0002) [reply tests/test_custom_checker.py:7]: Receive the following error when running tests/test_custom_checker.py with
```
from xmltodict import parse
from pyQuARC.code.custom_checker import CustomChecker

```
```
Traceback (most recent call last):
  File "/Users/sam/Documents/GitHub/pyQuARC/tests/test_custom_checker.py", line 2, in <module>
    from pyQuARC.code.custom_checker import CustomChecker
ModuleNotFoundError: No module named 'pyQuARC'
```

File: tests/test_schema_validator.py (hunk 9 above)
Comment (spa0002) [reply tests/test_schema_validator.py:3]: Receive the following error when running tests/test_schema_validator.py with
```
import os
from xmltodict import parse
from pyQuARC.code.schema_validator import SchemaValidator

```
```
Traceback (most recent call last):
  File "/Users/sam/Documents/GitHub/pyQuARC/tests/test_schema_validator.py", line 3, in <module>
    from pyQuARC.code.schema_validator import SchemaValidator
ModuleNotFoundError: No module named 'pyQuARC'
```

File: tests/test_string_validator.py (hunk 8 above)
Comment (spa0002) [reply tests/test_string_validator.py:3]: Receive the following error when running tests/test_custom_checker.py with
`from pyQuARC.code.string_validator import StringValidator`
```
Traceback (most recent call last):
  File "/Users/sam/Documents/GitHub/pyQuARC/tests/test_string_validator.py", line 1, in <module>
    from pyQuARC.code.string_validator import StringValidator
ModuleNotFoundError: No module named 'pyQuARC'

```

File: tests/test_downloader.py (hunk 10 above)
Comment (spa0002) [reply tests/test_downloader.py:3]: Receive the following error when running tests/test_downloader.py with
`from pyQuARC.code.downloader import Downloader`
```
Traceback (most recent call last):
  File "/Users/sam/Documents/GitHub/pyQuARC/tests/test_downloader.py", line 1, in <module>
    from pyQuARC.code.downloader import Downloader
ModuleNotFoundError: No module named 'pyQuARC'
```

File: tests/test_checker.py [hunk 15]
```diff
@@ -1,9 +1,9 @@
...
+sys.path.append(os.getcwd())
```
[the remark this answers is not in the document]
Comment (spa0002) [reply tests/test_checker.py:3]: Receive the following error message when running tests/test_checker.py with
```
from pyQuARC.code.checker import Checker

from .fixtures.checker import FUNCTION_MAPPING
from .common import read_test_metadata

```
```
Traceback (most recent call last):
  File "/Users/sam/Documents/GitHub/pyQuARC/tests/test_checker.py", line 1, in <module>
    from pyQuARC.code.checker import Checker
ModuleNotFoundError: No module named 'pyQuARC'
```

File: pyQuARC/main.py (hunk 3 above)
Comment (spa0002) [reply pyQuARC/main.py:10]: This appears to be a mistake. Lines 10 and 17 should be removed.

File: pyQuARC/main.py (hunk 4 above)
Comment (spa0002) [reply pyQuARC/main.py:14]: This appears to be a mistake. Lines 10 and 17 should be removed.

File: pyQuARC/code/gcmd_validator.py (hunk 2 above)
Comment (spa0002) [reply pyQuARC/code/gcmd_validator.py:5]: @svbagwell and I are looking into this. To be sure, what is meant by "'installed' package"?

File: tests/test_checker.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (xhagrg) [reply tests/test_checker.py:3]: This should not be the case when running `pytest tests/test_checker.py` from the root level.

File: pyQuARC/code/custom_validator.py (hunk 1 above)
Comment (xhagrg) [reply pyQuARC/code/custom_validator.py:183]: it might just be because of missing parenthesis. i.e. `if (description := url_obj.get(key)) in seen:`, see if that works?

File: pyQuARC/code/gcmd_validator.py (hunk 2 above)
Comment (xhagrg) [reply pyQuARC/code/gcmd_validator.py:5]: pyQuARC is a package. once you install it using `python setup.py install`, does it work the same in any python shell run outside of pyQuARC folder?

File: pyQuARC/main.py (hunk 4 above)
Comment (xhagrg) [reply pyQuARC/main.py:14]: This hasn't been fixed yet?

**Review by xhagrg — COMMENTED** [review]
Few small comments.

File: pyQuARC/code/gcmd_validator.py (hunk 2 above)
Comment (svbagwell) [reply pyQuARC/code/gcmd_validator.py:5]: this seems to be working for me, i did not get any errors

File: tests/test_string_validator.py [hunk 16]
```diff
@@ -1,6 +1,4 @@
 from pyQuARC.code.string_validator import StringValidator
-
-
```
Comment (xhagrg) [inline tests/test_string_validator.py:3]: Why remove the space between the import and classes?

**Review by xhagrg — COMMENTED** [review]
1. Some spaces between imports and class initializations are removed. Might be better to have them reverted. 
2. Was not sure about the `from pyQuARC.code.utils import get_headers` vs `from .utils import get_headers` need someone to help verify.

**Comment by xhagrg** [comment]
LGTM.
